apollo-router/src/plugins/rhai/mod.rs

```diff
@@ -512,6 +661,172 @@ macro_rules! gen_map_response {
+                            body.into(),
+                        )
+                        .into(),
+                    };
+                    let shared_response = Shared::new(Mutex::new(Some(response)));
+
+                    let result =
+                        execute(&$rhai_service, &$callback, (shared_response.clone(),));
+                    if let Err(error) = result {
+                        tracing::error!("map_response callback failed: {error}");
+                        let error_details = process_error(error);
+                        let response_opt = shared_response.lock().unwrap().take();
+                        return Ok(failure_message(
+                            response_opt.unwrap().context,
+                            error_details
+                        ));
+                    }
+
+                    let response_opt = shared_response.lock().unwrap().take();
+
+                    let $rhai_response { context, response } =
+                        response_opt.unwrap();
+                    let (parts, body) = http::Response::from(response).into_parts();
+
+                    let ctx = context.clone();
+
+                    let mapped_stream = rest
+                        .map_err(BoxError::from)
+                        .and_then(move |deferred_response| {
+                        let rhai_service = $rhai_service.clone();
+                        let context = ctx.clone();
+                        let callback = $callback.clone();
+                        async move {
+                            let response = $rhai_deferred_response {
+                                context,
+                                response: deferred_response.into(),
+                            };
+                            let shared_response = Shared::new(Mutex::new(Some(response)));
+
+                            let result = execute(
+                                &rhai_service,
+                                &callback,
+                                (shared_response.clone(),),
+                            );
+
+                            if let Err(error) = result {
+                                tracing::error!("map_response callback failed: {error}");
+                                let error_details = process_error(error);
+                                let error = Error {
+                                    message: error_details.message.unwrap_or_default(),
+                                    ..Default::default()
+                                };
+                                // We don't have a structured response to work with here. Let's
+                                // throw away our response and custom build an error response
+                                let error_response = graphql::Response::builder()
+                                    .errors(vec![error]).build();
+                                return Ok(serde_json::to_vec(&error_response)?.into());
```

Comment:
Should we just send an error payload or drop the stream alltogether when this happens? This might have small implications in a defer context, but bigger ones in a subscriptions one, so I'm not sure.